pennylane_qiskit/qiskit_device.py

```diff
@@ -267,258 +314,27 @@ def _get_state(self, result):
-            return (eigvals @ prob).real
-
-        if self.analytic:
-            # Raise a warning if backend is a hardware simulator
-            warnings.warn(self.hw_analytic_warning_message.format(self.backend), UserWarning)
-
-        # estimate the ev
-        return np.mean(self.sample(observable, wires, par))
-
-    def var(self, observable, wires, par):
-        if self.backend_name in self._state_backends and self.analytic:
-            # exact variance value
-            eigvals = self.eigvals(observable, wires, par)
-            prob = np.fromiter(self.probability(wires=wires).values(), dtype=np.float64)
-            return (eigvals ** 2) @ prob - (eigvals @ prob).real ** 2
-
-        if self.analytic:
-            # Raise a warning if backend is a hardware simulator
-            warnings.warn(self.hw_analytic_warning_message.format(self.backend), UserWarning)
-
-        return np.var(self.sample(observable, wires, par))
-
-    def sample(self, observable, wires, par):
-        if observable == "Identity":
-            return np.ones([self.shots])
+    def generate_samples(self):
 
         # branch out depending on the type of backend
         if self.backend_name in self._state_backends:
-            # software simulator. Need to sample from probabilities.
-            eigvals = self.eigvals(observable, wires, par)
-            prob = np.fromiter(self.probability(wires=wires).values(), dtype=np.float64)
-            return np.random.choice(eigvals, self.shots, p=prob)
-
-        # a hardware simulator
-        if self.memory:
-            # get the samples
-            samples = self._current_job.result().get_memory()
-
-            # reverse qubit order to match PennyLane convention
-            samples = np.vstack([np.array([int(i) for i in s[::-1]]) for s in samples])
-
-        else:
-            # Need to convert counts into samples
-            samples = np.vstack(
-                [np.vstack([s] * int(self.shots * p)) for s, p in self.probability().items()]
-            )
-
-        if isinstance(observable, str) and observable in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
-            return 1 - 2 * samples[:, wires[0]]
-
-        eigvals = self.eigvals(observable, wires, par)
-        wires = np.hstack(wires)
-        res = samples[:, np.array(wires)]
-        samples = np.zeros([self.shots])
+            # software simulator: need to sample from probabilities
+            return super().generate_samples()
 
-        for w, b in zip(eigvals, itertools.product([0, 1], repeat=len(wires))):
-            samples = np.where(np.all(res == b, axis=1), w, samples)
+        # hardware or hardware simulator
+        samples = self._current_job.result().get_memory()
 
-        return samples
+        # reverse qubit order to match PennyLane convention
+        return np.vstack([np.array([int(i) for i in s[::-1]]) for s in samples])
 
     @property
     def state(self):
         return self._state
```

Comment:
Will there be issues with qubit ordering here, due to the difference in conventions between PL and Qiskit? If so, this might need to be fixed to produce correct analytic probability.

Reply:
Oh right! So the logic of this part was not changed, this was part of `sample`. The state representation is reversed by having `s[::-1]`. Is this what you mean? The comment could be updated if it is misleading.

Reply:
An example:
```python
from qiskit import QuantumCircuit, execute
from qiskit import Aer, IBMQ
from qiskit.compiler import assemble, transpile
import numpy as np

num_qubits = 2
circ = QuantumCircuit(2, 2)
circ.h(1)
circ.measure([0,1], [0,1])

backend = Aer.get_backend('qasm_simulator')

compiled_circuits = transpile(circ, backend=backend)
qobj = assemble(experiments=compiled_circuits, backend=backend,
                             shots=10, memory=True)
job = backend.run(qobj)
result = job.result()

samples = result.get_memory(0)
print('Qiskit samples: ', samples)

pennylane_samples = np.vstack([np.array([int(i) for i in s[::-1]]) for s in samples])
print('PennyLane samples: \n', pennylane_samples)
```
Which outputs:
```python
Qiskit samples:  ['10', '00', '10', '10', '00', '00', '00', '10', '10', '10']
PennyLane samples: 
 [[0 1]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 1]]
```

Reply:
Oh, I was trying to comment on the 
```python
    def state(self):
        return self._state
```
part! GitHub tries to show the closest change though 😆 

I wasn't 100% sure about this, because I saw the old version of the device didn't do any processing here either, but I figured if nothing is being done, the user will receive the statevector in the Qiskit convention?

Reply:
But you answered my question below 🙂